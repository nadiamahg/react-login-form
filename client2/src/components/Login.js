import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { login } from "../actions/authActions";
import image from '../images/img-01.png';
import Tilt from 'react-tilt'
export class Login extends React.Component {
    constructor() {
        super();
        this.state = {
            email: "",
            password: "",
            errors: {}
        };
    }
    componentDidMount() {
      // If logged in and user navigates to Login page, should redirect them to dashboard
      if (this.props.auth.isAuthenticated) {
        this.props.history.push("/dashboard");
      }
    }
    componentWillReceiveProps(nextProps) {
      if (nextProps.auth.isAuthenticated) {
        this.props.history.push("/dashboard"); // Redirect to dashboard when they log in
      }
      if (nextProps.errors) {
        this.setState({
          errors: nextProps.errors
        });
      }
    }
    onChange = e => {
        this.setState({
            [e.target.id]: e.target.value
        });
    };
    onSubmit = e => {
        e.preventDefault();
        const userData = {
            email: this.state.email,
            password: this.state.password
        };
        this.props.login(userData);
    };
    render() {
        const { errors } = this.state;
        return (
          <div className="limiter">
            <div className="page-container">
              <div className="wrap">
                <div className="login-pic">
                  <Tilt className="Tilt dashboard-img" options={{ max : 25 }} style={{ height: 250, width: 250 }} >
                    <div className="Tilt-inner"> <img src={image} alt="IMG"/></div>
                  </Tilt>
                </div>
                <form className="login-form" onSubmit={this.onSubmit}>
                  <span className="page-title">
                    Login
                  </span>
                  <div className="wrap-input">
                    <input onChange={this.onChange} value={this.state.email} className="input" type="text" id="email" placeholder="Email"/>
                    <div className="login-error red-text" id="emailError">{errors.email}</div>
                  </div>
                  <div className="wrap-input">
                    <input onChange={this.onChange} value={this.state.password} className="input" type="password" id="password" placeholder="Password"/>
                    <div className="login-error red-text" id="passwordError">{errors.password}</div>
                  </div>
                  <div className="container-btn">
                    <button type="submit" className="page-btn">
                      Login
                    </button>
                  </div>
                </form>
              </div>
            </div>
          </div>
        );
    }
}
Login.propTypes = {
  login: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired
};
const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});
export default connect(
  mapStateToProps,
  { login }
)(Login);