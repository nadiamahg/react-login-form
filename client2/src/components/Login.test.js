import React from "react";
import { mount, configure, shallow } from "enzyme";
import {Login} from "./Login";
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

const mockLoginfn = jest.fn();
const props = {
    auth : {
        isAuthenticated: false
    },
    errors: {},
    login: mockLoginfn
}

const mockHistory = { push: jest.fn() };

configure({ adapter: new Adapter() })

describe("Login Component", () => {
     const wrapper = mount(<Login {...props} history={mockHistory}/>)

     it('should render Login correctly', () => {
        const component = renderer.create(<Login {...props}/>,);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should render form without throwing an error', () => {
        expect(wrapper.find('form').exists()).toBe(true);
    })
    
    it('should have a login button', ()=> {
        expect(wrapper.find('button').text()).toEqual('Login');
        expect(wrapper.find('button').prop('type')).toEqual('submit');
    });

    it('should have input fields for email and password', ()=> {
        expect(wrapper.find('input#email')).toHaveLength(1);
        expect(wrapper.find('input#password')).toHaveLength(1);
    });

    it('should call the mock login function', () => {
       wrapper.find('form').simulate('submit', {preventDefault() {}});
       expect(mockLoginfn.mock.calls.length).toBe(1);
    })

    it('should have an empty email and password state var', ()=> {
        expect(wrapper.state('email')).toEqual('');
        expect(wrapper.state('password')).toEqual('');
    });

    it('should navigate to dashboard page when user is authenticated', ()=> {
        const newProps = {
            auth : {
                isAuthenticated: true
            },
            errors: {},
            login: mockLoginfn
        }
        const wrapper = mount(<Login {...newProps} history={mockHistory} />);
        expect(mockHistory.push.mock.calls[0]).toEqual(['/dashboard']);
    });

    it('should have validation error messages when there are errors for email and password', ()=> {
        const errors = {
            email: "Email error",
            password: "Password error"
        }
        wrapper.setState({errors});
        expect(wrapper.find('div#emailError').text()).toEqual('Email error');
        expect(wrapper.find('div#passwordError').text()).toEqual('Password error');

    });

    it('should navigate to dashboard page when user is authenticated in nextProps', ()=> {
        wrapper.setProps({auth: {isAuthenticated: true}})
        expect(mockHistory.push.mock.calls[0]).toEqual(['/dashboard']);
    });
});