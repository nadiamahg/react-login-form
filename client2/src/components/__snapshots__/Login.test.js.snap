// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should render Login correctly 1`] = `
<div
  className="limiter"
>
  <div
    className="page-container"
  >
    <div
      className="wrap"
    >
      <div
        className="login-pic"
      >
        <div
          className="Tilt dashboard-img"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          style={
            Object {
              "height": 250,
              "width": 250,
            }
          }
        >
          <div
            className="Tilt-inner"
          >
             
            <img
              alt="IMG"
              src="img-01.png"
            />
          </div>
        </div>
      </div>
      <form
        className="login-form"
        onSubmit={[Function]}
      >
        <span
          className="page-title"
        >
          Login
        </span>
        <div
          className="wrap-input"
        >
          <input
            className="input"
            id="email"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value=""
          />
          <div
            className="login-error red-text"
            id="emailError"
          />
        </div>
        <div
          className="wrap-input"
        >
          <input
            className="input"
            id="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
          <div
            className="login-error red-text"
            id="passwordError"
          />
        </div>
        <div
          className="container-btn"
        >
          <button
            className="page-btn"
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
